### PR TITLE
feat4: filter functionality

### PR DESCRIPTION
**Pull request by SinchanaMS**
Added the filter functionality to the Products Listing Page.

1. Sort by Price
2. Price Range Slider
3. Category Filters
4. Sort by Rating
5. Reset filters on Clear

You may ignore the CSS files as they were already reviewed in the first submission.

Preview:

https://user-images.githubusercontent.com/94738532/160301686-05b3ff61-7543-4582-931e-897d18e4f84a.mp4

### CODE REVIEW

File: src/contexts/FilterContext.jsx [hunk 1]
```diff
@@ -0,0 +1,25 @@
+import { createContext, useContext, useReducer, useState, useEffect } from "react";
+import { filterReducer } from "../reducer/filterReducer";
+
```
Comment (dev-ashishsethi) [inline src/contexts/FilterContext.jsx:3]: I don't think context was needed for filters if you are filtering only on single page

File: src/components/FilterPane/FilterCompose.jsx [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+    },products)
+}
+
+export const getProductList =(filterState, products)=> applyFilters(
+    filterState,
+    sortByPrice,
+    sortByRating,
+    filterByCategory,
+    filterByPriceRange
+)(products)
```
Comment (dev-ashishsethi) [inline src/components/FilterPane/FilterCompose.jsx]: good use of iife

**Review by dev-ashishsethi — APPROVED** [review]
do check comments

File: src/components/FilterPane/FilterCompose.jsx [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+            return sortedList.sort((prod1, prod2) => prod2.price.discounted - prod1.price.discounted)
+        case "LOW-TO-HIGH":
+            return sortedList.sort((prod1, prod2) => prod1.price.discounted - prod2.price.discounted)
+        default:
+            return sortedList
+    }
+}
+
+const sortByRating = ({rating}, products) => {
+    return products.filter(product => product.rating >= rating)
```
Comment (pankaj-prog) [inline src/components/FilterPane/FilterCompose.jsx]: it should be filter by instead of sort by

File: src/reducer/filterReducer.jsx [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+        category:[],
+        rating: "",
+        range: 200
+    }
+    switch (action.type){
+        case "SORT_BY":
+            return {...filterState, sortby: action.payload}
+        case "CATEGORY":
+            const {category} = filterState
+          return  category.includes(action.payload) ? {...filterState, category: category.filter(category => category!==action.payload)} :
```
Comment (pankaj-prog) [inline src/reducer/filterReducer.jsx:13]: I like this, the way have used same type for both checking and unchecking, gonna implement this in my project as well

**Review by pankaj-prog — APPROVED** [review]

File: src/contexts/FilterContext.jsx (hunk 1 above)
Comment (SinchanaMS) [reply src/contexts/FilterContext.jsx:3]: Thank you, I'll check.

File: src/components/FilterPane/FilterCompose.jsx (hunk 3 above)
Comment (SinchanaMS) [reply src/components/FilterPane/FilterCompose.jsx]: Oh right! Thanks for pointing it out.

**Comment by SinchanaMS** [comment]
Thank you for the reviews, I have made the PR changes. Merging to dev.

**Pull request merged**